osgp/platform/osgp-adapter-domain-core/src/main/java/org/opensmartgridplatform/adapter/domain/core/application/services/FirmwareManagementService.java

```diff
@@ -325,10 +333,103 @@ private void checkFirmwareHistory(final String deviceId,
     }
 
+    /**
+     * @param deviceId
+     *            the id of the device we are checking
+     * @param firmwareVersions
+     *            the list of firmware modules versions (so type and version) to
+     *            check if they are currently installed on the device, using the
+     *            history of the devices firmware history
+     * @return a list of firmware module versions not present in the the devices
+     *         firmware history
+     * @throws FunctionalException
+     */
+    public List<FirmwareVersion> checkFirmwareHistoryForModuleVersionsNotCurrentlyInstalled(final String deviceId,
+            final List<FirmwareVersion> firmwareVersions) {
+
+        if (firmwareVersions.isEmpty()) {
+            return firmwareVersions;
+        }
+        // copy input parameter
+        final List<FirmwareVersion> firmwareVersionsToCheck = new ArrayList<>();
+        firmwareVersionsToCheck.addAll(firmwareVersions);
+
+        // gets list of all historically installed modules
+        final Device device = this.deviceRepository.findByDeviceIdentification(deviceId);
+        final List<DeviceFirmwareFile> deviceFirmwareFiles = this.deviceFirmwareFileRepository
+                .findByDeviceOrderByInstallationDateAsc(device);
+
+        // Next section can be removed
+        final List<FirmwareVersion> firmwareVersionsInHistory = deviceFirmwareFiles.stream()
+                .map(d -> d.getFirmwareFile().getModuleVersions().entrySet())
+                .flatMap(Collection::stream)
+                .map(e -> new FirmwareVersion(FirmwareModuleType.forDescription(e.getKey().getDescription()),
+                        e.getValue()))
+                .collect(Collectors.toList());
+
+        // Transform this list so it contains only the latest entry for each
+        // moduleType, using
+        // firmwareVersionsInHistory[x].DeviceFimwareFile.firmwareFile.firmwareModules.firmwareModule.description.
+        // and firmwareVersionsInHistory[x].installationDate
+        Map<String, FirmwareVersionWithInstallationDate> currentlyInstalledFirmwareVersionsPerType = new HashMap<String, FirmwareVersionWithInstallationDate>();
+
+        for (DeviceFirmwareFile firmwareFile : deviceFirmwareFiles) {
+
+            Map<FirmwareModule, String> fwms = firmwareFile.getFirmwareFile().getModuleVersions(); // .keySet();
+            Date installationDate = firmwareFile.getInstallationDate();
+
+            for (Map.Entry<FirmwareModule, String> entry : fwms.entrySet()) {
+                String version = entry.getValue();
+                FirmwareModule fwm = entry.getKey();
+                if (currentlyInstalledFirmwareVersionsPerType.containsKey(fwm.getDescription())) {
+                    // check if this installation of this same kind of module is
+                    // of a later date
+                    if (currentlyInstalledFirmwareVersionsPerType.get(fwm.getDescription()).installationDate
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS134)